mailpoet/tests/integration/Automation/Integrations/MailPoet/Actions/SendEmailActionTest.php

```diff
@@ -52,6 +56,8 @@ class SendEmailActionTest extends \MailPoetTest {
 
   public function _before() {
     parent::_before();
+    $this->cleanup();
```

Comment:
Since I was tagged I can share my opinion, but you can do whatever you think is better since I don't code much anymore.
1. If tests rely on emptying the database and we clean in the `_after` method we get flakey tests. Sometimes people forget to clean up properly and tests start failing when running the whole suite but pass if you run a single test. It is a pain to debug this kind of failures. And if that depends on the execution order, it is even more challenging to debug. So the tests are more stable if they don't expect anything and prepare the environment for them. 
2. If clean in both `_after` and `_before` we waste time. Running the cleaning queries takes time and if it is done in all tests it adds up and makes the tests run longer. 

So that is why I prefer if tests run the clean up before the execution.

Reply:
> If clean in both _after and _before we waste time. Running the cleaning queries takes time, and if it is done in all tests it adds up and makes the tests run longer.

Yes. But as we are inconsistent in the approach adding to both `before` and `after` is the safest variant. The cleanup in the before section protects the test against ones that forget to cleanup (or were written only with cleanup in before), and adding cleanup in the after method protects tests that don't do cleanup in the before section.

So I guess this is something that needs to be discussed on the team level (I mean MailPoet team), and we need to unify the approach in existing tests. Until we use the same approach in all tests I would keep adding it to both.

Reply:
I'm ok with adding the cleanup() to both _before() and _after(), but I don't think it is needed in this case as, if I'm remembering correctly a check that I did last week, the test suite passes if we only call cleanup() inside _after().

> It should be the responsibility of the test to prepare the environment for its run. This applies to both preparing testing data and cleaning data in DB tables it is about to use.

I have a different view of the conventions regarding tests, but I'm talking off the top of my head so I might be misremembering things. I always saw things the opposite way as what was described above. Individual tests should make sure that they finish leaving the environment in the same state that they found it when they started.

> doing the cleanup after the run feels more natural to me (I clean my mess)

Same here

> Doing the cleanup after the test implicates interdependence between tests.

I also don't agree with this point.

> If tests rely on emptying the database and we clean in the _after method we get flakey tests.

I wouldn't call them flaky as they consistently fail or pass depending on whether they are being triggered individually or with the whole suite.

> Sometimes people forget to clean up properly and tests start failing when running the whole suite but pass if you run a single test. It is a pain to debug this kind of failures. And if that depends on the execution order, it is even more challenging to debug.

That is a fair point. In my opinion, we should address this by fixing the tests that are not cleaning up properly. I have fixed a few already by creating a MailPoetTest::_after() method that prints the number of entries in the affect tables and then running the whole test suite. This way it is not that hard to find the tests that are leaving entries in the database. But maybe there are other instances where it is not that easy to find the problematic tests for some reason?

As a side note, WP runs all the queries for a particular test inside a single transaction that is never committed. This way data is never written to the database and the tests run much faster. I don't know if we could do something similar for our tests.

Reply:
I added the cleanup to the after method. So now we do it both before and after test.

> But maybe there are other instances where it is not that easy to find the problematic tests for some reason?

There are some tests that create WordPress users or add some hooks. Forgotten WP users may easily cause failing tests as a lot of them to work with subscribers, and WP users are automatically added to the WordPress users list.